### PR TITLE
feat: taker fee should be passed to txfees module

### PR DESCRIPTION
**Pull request by mtsitrin**
This PR simplifies the taker fee logic once the txfee module is available.

taker fee checked vs the supported fees on txfee module.
if not supported, it will be swapped on the first pool on the route (to get a whitelisted pool asset)
than it sent to the txfee module to be burned

### CODE REVIEW

File: x/poolmanager/types/expected_keepers.go
Comment (mtsitrin) [inline x/poolmanager/types/expected_keepers.go:1]: extracted from [x/poolmanager/types/routes.go](https://github.com/dymensionxyz/osmosis/pull/32/files#diff-f64b652d2aa53de44319c25b808783a312503622c02a29cdc2687579dfdb20a2)

File: x/poolmanager/types/routes.go [hunk 1]
```diff
@@ -2,69 +2,12 @@ package types
...
-	// CalcInAmtGivenOut calculates the amount of tokenIn given tokenOut and the pool's current state.
-	// Returns error if the given pool is not a CFMM pool. Returns error on internal calculations.
-	CalcInAmtGivenOut(
-		ctx sdk.Context,
-		poolI PoolI,
-		tokenOut sdk.Coin,
-		tokenInDenom string,
-		swapFee sdk.Dec,
-	) (tokenIn sdk.Coin, err error)
+type SwapRoutesUnion struct {
```
Comment (mtsitrin) [inline x/poolmanager/types/routes.go]: union struct so we could have common API for taker fee from both `SwapIn` and `SwapOut`

File: x/gamm/keeper/keeper.go [hunk 2]
```diff
@@ -73,10 +85,18 @@ func (k *Keeper) SetHooks(gh types.GammHooks) *Keeper {
...
 
+// SetPoolManager sets the pool manager.
+// must be called when initializing the keeper.
 func (k *Keeper) SetPoolManager(poolManager types.PoolManager) {
 	k.poolManager = poolManager
 }
 
+// SetTxFees sets the tx fees keeper.
+// must be called when initializing the keeper.
+func (k *Keeper) SetTxFees(txfees types.TxFeeKeeper) {
```
Comment (omritoptix) [inline x/gamm/keeper/keeper.go:96]: Why is it not being initialized when we init the keeper?

File: x/gamm/keeper/msg_server.go [hunk 3]
```diff
@@ -178,7 +179,12 @@ func (server msgServer) SwapExactAmountIn(goCtx context.Context, msg *types.MsgS
 		return nil, err
 	}
 
-	err = server.keeper.chargeTakerFeeSwapAmountIn(ctx, takerFeesCoins, sender, msg.Routes)
+	routesForTakerFee := poolmanagertypes.SwapRoutesUnion{
+		InRoutes:  msg.Routes,
+		OutRoutes: nil,
+		Type:      "in",
+	}
+	err = server.keeper.chargeTakerFee(ctx, takerFeesCoins, sender, routesForTakerFee, msg.Routes[len(msg.Routes)-1].TokenOutDenom)
```
Comment (omritoptix) [inline x/gamm/keeper/msg_server.go]: isn't it supposed to be msg.Routes[0].TokenOutDenom to get the out token of the first people given that it's an "in" route and this `msg.Routes[len(msg.Routes)-1]` should be for the Out route?

File: x/gamm/keeper/keeper.go (hunk 2 above)
Comment (mtsitrin) [reply x/gamm/keeper/keeper.go:96]: theres a dependecy cycle
someone needs to be initialized first

**Pull request merged**